adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java

```diff
@@ -1226,11 +1185,75 @@ protected final void onCommandReceived(
+    private static void addRetainAnnotation(final MqttContext context, final Message downstreamMessage,
+            final Span currentSpan) {
 
         if (context.message().isRetain()) {
             currentSpan.log("device wants to retain message");
             MessageHelper.addAnnotation(downstreamMessage, MessageHelper.ANNOTATION_X_OPT_RETAIN, Boolean.TRUE);
         }
     }
+
+    private Future<TenantObject> isAdapterEnabled(final TenantObject tenantConfig) {
+        if (tenantConfig.isAdapterEnabled(getTypeName())) {
+            LOG.debug("protocol adapter [{}] is enabled for tenant [{}]",
+                    getTypeName(), tenantConfig.getTenantId());
+            return Future.succeededFuture(tenantConfig);
+        } else {
+            LOG.debug("protocol adapter [{}] is disabled for tenant [{}]",
+                    getTypeName(), tenantConfig.getTenantId());
+            return Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_FORBIDDEN,
+                    "adapter disabled for tenant"));
+        }
+    }
+
+    private Future<DeviceUser> authenticateCredentials(final DeviceCredentials credentials) {
+        final Future<DeviceUser> result = Future.future();
+        usernamePasswordAuthProvider.authenticate(credentials, result.completer());
+        return result;
+    }
+
+    private Future<Device> isDeviceRegistered(final Device authenticatedDevice, final DeviceCredentials credentials,
+            final Span currentSpan) {
+        currentSpan.log(String.format("device authenticated"));
+        LOG.debug("successfully authenticated device [tenant-id: {}, auth-id: {}, device-id: {}]",
+                authenticatedDevice.getTenantId(), credentials.getAuthId(),
+                authenticatedDevice.getDeviceId());
+        return getRegistrationAssertion(authenticatedDevice.getTenantId(),
+                authenticatedDevice.getDeviceId(),
+                authenticatedDevice, currentSpan.context())
+                        .compose(ok -> {
+                            currentSpan.log(String.format("device registration verified"));
+                            LOG.debug("verified device registration [tenant-id: {}, device-id: {}]",
+                                    authenticatedDevice.getTenantId(), authenticatedDevice.getDeviceId());
+                            return accepted(authenticatedDevice);
+                        });
+    }
+
+    private Future<Device> createLink(final MqttEndpoint endpoint, final Device authenticatedDevice,
+            final Span currentSpan) {
+        final Future<MessageSender> telemetrySender = getTelemetrySender(authenticatedDevice.getTenantId());
+        final Future<MessageSender> eventSender = getEventSender(authenticatedDevice.getTenantId());
+
+        return CompositeFuture
+                .all(getRegistrationClient(authenticatedDevice.getTenantId()), telemetrySender, eventSender)
```

Comment:
I believe that at this point we already have created a registration client because we have already verified that the device exists and is enabled, right? It shouldn't have too much of an impact because the client will be retrieved from the cache. However, for clarity I think you should remove it from here and just create the senders.

Reply:
Done